### ISSUE TITLE
non-linear memory usage for wide data

### ISSUE THREAD

**Issue by arnocandel** [dmlc/xgboost#6759]
```
from xgboost import XGBRegressor
import numpy as np
import time

for cols in [500000, 600000]:
    shape = (1000, cols)
    print("shape: %s" % str(shape))

    X = np.random.rand(*shape)
    y = np.random.rand(shape[0])

    # GPU
    t = time.time()
    try:
        model = XGBRegressor(tree_method='gpu_hist', n_estimators=1, max_bin=64)
        model.fit(X, y)
        print("GPU fit took: %s" % (time.time() - t))
    except:
        print("OOM")
```

on RTX 2080Ti 11GB:
```
shape: (1000, 500000)
GPU fit took: 11.824267148971558
shape: (1000, 600000)
OOM
```

on RTX 3090 24GB:
```
shape: (1000, 500000)
GPU fit took: 8.36401629447937
shape: (1000, 600000)
OOM
```

One would expect that 20% more data would fit into more than 2x memory.

Also, sometimes RTX 3090 takes up to 60s for same operation, idle system.
```
arno@rippa:/data$ python xgb-wide.py 
shape: (1000, 500000)
GPU fit took: 60.621460914611816
shape: (1000, 600000)
OOM
arno@rippa:/data$ python xgb-wide.py 
shape: (1000, 500000)
GPU fit took: 12.049131631851196
shape: (1000, 600000)
OOM
arno@rippa:/data$ python xgb-wide.py 
shape: (1000, 500000)
GPU fit took: 7.468593120574951
shape: (1000, 600000)
OOM
```

tested with latest nightly:
https://s3-us-west-2.amazonaws.com/xgboost-nightly-builds/xgboost-1.4.0_SNAPSHOT%2B4f75f514ce48936e654c6baddccb7fb864e16af2-py3-none-manylinux2010_x86_64.whl

**Comment by trivialfis** [comment]
The internal data representation of xgboost is a CSR array, so the memory usage would be high for wide datasets.  We might be able to ingest CPU data directly (bypassing the CSR array) for training in the future and do more optimization for this kind of dataset.

**Comment by arnocandel** [comment]
thanks - any suggestions how to avoid the non-linearity (500k cols fits into 11GB card, 600k cols doesn't fit into 24GB card)?

**Comment by trivialfis** [comment]
I will run some tests tomorrow.

**Comment by trivialfis** [comment]
Quick note:  The peak memory usage is in sketching where sort requires a double buffer.  We batch over it based on the memory requirement.  But now the calculation of that requirement is a bit outdated.

**Comment by andyatkinson20** [comment]
> thanks - any suggestions how to avoid the non-linearity (500k cols fits into 11GB card, 600k cols doesn't fit into 24GB card)?

This is happening for me as well.  After investing in a 3090 I was hoping for a more complex model to fit in memory.  I'm working on R and so far haven't found the DeviceQuantileDMatrix option, which might help?

**Comment by trivialfis** [comment]
We will make a summary when we do the release, but for anyone curious about the reason for closing this issue, here are the recent optimizations:
- Further compression for the gradient index.
- Multi-level quantile sketching.
- A new option to limit the size of the quantile sketching container.
- Improved integration with RMM.
- Improved support for iterator (QuantileDMatrix).

The root cause of this issue is the `sort` in the quantile sketching, which is unlikely to go away, but the above mitigations should be able to provide more control.
